File: 0023-merge-k-sorted-lists/0023-merge-k-sorted-lists.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    struct compare{
        bool operator()(ListNode* a, ListNode* b){
            return a->val>b->val;
        }  
    };
    ListNode* mergeKLists(vector<ListNode*>& lists) {
        priority_queue<ListNode*, vector<ListNode*>, compare> minHeap;
        for(ListNode* element: lists){
            if(element!=NULL){
                minHeap.push(element);
            }
        }
        ListNode* head=NULL;
        ListNode* tail=NULL;
        while(!minHeap.empty()){
            ListNode* curr = minHeap.top();
            minHeap.pop();
            if(head==NULL){
                head=curr;
                tail=curr;
            }
            else{
                tail->next=curr;
                tail=curr;
            }
            if(curr->next!=NULL){
                minHeap.push(curr->next);
            }
        }
        return head;
    }
};

File: 0023-merge-k-sorted-lists/0023-merge-k-sorted-lists.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode mergeKLists(ListNode[] lists) {
        PriorityQueue<ListNode> minHeap = new PriorityQueue<>((a, b)-> a.val - b.val);
        for(ListNode element: lists){
            if(element!=null){
                minHeap.add(element);
            }
        }
        ListNode head=null;
        ListNode tail=null;
        while(!minHeap.isEmpty()){
            ListNode curr = minHeap.poll();
            if(head==null){
                head = curr;
                tail = curr;
            }
            else{
                tail.next = curr;
                tail = curr;
            }
            if(curr.next!=null){
                minHeap.add(curr.next);
            }
        }
        return head;
    }
}